src/v/cloud_topics/dl_stm/dl_stm_state.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "cloud_topics/dl_overlay.h"
+#include "cloud_topics/dl_stm/dl_version.h"
+#include "container/fragmented_vector.h"
+
+namespace experimental::cloud_topics {
+
+struct dl_overlay_entry {
+    dl_overlay overlay;
+
+    dl_version added_at;
+    dl_version removed_at;
+};
+
+/// In-memory state of the data layout state machine (dl_stm).
+///
+/// Separating the state from the state machine allows the state to be
+/// checkpointed and restored independently of the state machine.
+class dl_stm_state {
+    friend class dl_stm_state_accessor;
+
+public:
+    /// Add a new overlay to the state. The overlay becomes visible
+    /// starting with the current version.
+    void push_overlay(dl_overlay overlay);
+
+    /// Find an overlay that contains the given offset. If no overlay
+    /// contains the offset, find the overlay covering the next closest
+    /// available offset.
+    std::optional<dl_overlay> lower_bound(kafka::offset offset) const;
+
+    void set_current_version(dl_version version) {
```

Comment:
It I remember the doc correctly the versioning was supposed to be done using added_at and removed_at fields but now we have an explicit version which is not a log offset. Is there any new design bits that I'm missing?

Reply:
Do you get this impression because of the existence of the `set_current_version` method? Because of existence dl_version named type?

The only intended divergence from the design is using batch base_offset rather than record offset to mitigate the potential problems caused by unexpected retries.

Do you think it would be better to extend push_overlay to accept 2 params, 1) dl_overlay and 2) dl_version at which the overlay is being added to avoid having this method which you need to remember to call?

Reply:
Btw, using batch base_offset rather than record offset is just a quick workaround of the problem (noted in code comment). Happy to discuss and change the approach there.

Reply:
Reworked this by making push_overlay (and all future state mutations will have to do the same, i.e. create_snapshot) accept dl_version at which mutation happens as a parameter. This should be less error prone.

Ideally, the api would look something like 

`start_transaction(dl_version).push_overlay(o).commit()` but i don't think so much abstraction is warranted. at least for now